containers/src/Data/Map/Internal.hs

```diff
@@ -4284,41 +4310,69 @@ bin k x l r
 
 
 {--------------------------------------------------------------------
-  Eq converts the tree to a list. In a lazy setting, this
-  actually seems one of the faster methods to compare two trees
-  and it is certainly the simplest :-)
+  Eq
 --------------------------------------------------------------------}
+
 instance (Eq k,Eq a) => Eq (Map k a) where
-  t1 == t2  = (size t1 == size t2) && (toAscList t1 == toAscList t2)
+  m1 == m2 = liftEq2 (==) (==) m1 m2
+  {-# INLINABLE (==) #-}
 
-{--------------------------------------------------------------------
-  Ord
---------------------------------------------------------------------}
+-- | @since 0.5.9
+instance Eq k => Eq1 (Map k) where
+  liftEq = liftEq2 (==)
+  {-# INLINE liftEq #-}
 
-instance (Ord k, Ord v) => Ord (Map k v) where
-    compare m1 m2 = compare (toAscList m1) (toAscList m2)
+-- | @since 0.5.9
+instance Eq2 Map where
+  liftEq2 keq eq m1 m2 = size m1 == size m2 && sameSizeLiftEq2 keq eq m1 m2
+  {-# INLINE liftEq2 #-}
+
+-- Assumes the maps are of equal size to skip the final check
```

Comment:
Is this `EqM` business better than just using `foldr`? If so, why?

Reply:
Yes it is, you can see a `foldr` version in #1016 (named `setCmp`).

To expand it out, `foldMap` with `EqM` becomes something like

```hs
                              -- v result of comparing the map and the iterator
go :: Map k a -> Stack k a -> (# Ordering, Stack k a #)
                                        -- ^ remaining iterator
```
We do this recursively and short-circuit on EQ.

In contrast, `foldr` becomes something like this, which is more of a continuation style

```hs
                                                      -- v final result
go :: Map k a -> Stack k a -> (Stack k a -> Ordering) -> Ordering
                             -- ^ if not EQ, call this with remaining iterator
```

Which is slower.

Reply:
All right, the slowness of the `foldr` should be documented if it isn't already. Do you know why it's slower? Is there a problem inlining `foldr` properly? Or an issue with arity analysis (does `oneShot` make a difference)?

Reply:
`oneShot` does improve things but it's still slower. Arity analysis is also working (I didn't try switching it off but it can only be worse).  
I'm not surprised that it is slower though. Not an expert here, but it has to heap allocate those continuation closures which is surely expensive. No allocations required in the foldMap version.

Reply:
Hmm? Why do we need to heap-allocate continuation closures when arity analysis is working? Isn't the state all in the `Stack`? I admit I'm not fully alert right now (I was kept up much of the night by intermittent screaming ☹️), so I could be missing something obvious.

Reply:
So we have in `Set.foldr`

```hs
go :: b -> Set a -> b
go z' (Bin _ x l r) = go (f x (go z' r)) l
```

`b` is `Stack a -> Ordering`, so

```hs
go :: (Stack a -> Ordering) -> Set a -> Stack a -> Ordering
go z' (Bin _ x l r) = go (f x (go z' r)) l
```

When arity analysis decides to eta-expand, we get

```hs
go :: (Stack a -> Ordering) -> Set a -> Stack a -> Ordering
go z' (Bin _ x l r) eta = go (f x (go z' r)) l eta
```

This has helped avoid creating and returning a closure, but there is still a closure as input.  
I don't know whether some transformation can get rid of that, if it's even possible.

---

This doesn't happen for something like `[]`, where `foldr` is

```hs
go :: [a] -> b
go (x : xs) = f x (go xs)
```

since there is no input `b` which sticks around.

Reply:
Ah, okay. I think I get that now.